### DISCUSSION TITLE
Correct way to type query factories?

### DISCUSSION THREAD

**Discussion by anddoutoi** [TanStack/query#2393] [Q&A]
I'm trying to wrap my head around how and where I should put the types when using _query factories_. I created a sandbox that somewhat shows how we use query factories today but I fail on TypeScript here.

Sandbox includes a simple hook that tried to infer type from the `fetchTodos()` that does not work. Another hook with a few types that does not work and a 3rd working hook that is ludicrous.

https://codesandbox.io/s/react-query-factories-421nb?file=/src/todos.tsx

PS. I also tried different constellations where I put the generics on `useQuery` instead of using `UseQueryOptions` but didn't get that to work.

**Comment by TkDodo** [comment]
you need the 3rd generic and need to pass it through, like so: https://codesandbox.io/s/react-query-factories-forked-pnl2l

**Reply by anddoutoi** [reply]
That makes the `select()` work for `useTypedGetTodosQuery()` but if you look at line 18 and 26 in `todos.tsx`, data is now `unknown`.

**Reply by TkDodo** [reply]
well yeah I forgot to carry along the default value for the generic, so `T` must become `T = Todo[]` in `useTypedGetTodosQuery`. I updated my sandbox above and it should work now.
